### PR TITLE
Feature/banner image improvements

### PR DESCRIPTION
**Pull request by dlipeles**
PR for item #1267 

Users need to be able to crop and scale their uploaded images to fit them to the tipping banner.

eV has approved React-Avatar-Editor, we will need to make some UI changes to support the resizing window and scaling slider.

### CODE REVIEW

File: app/javascript/packs/banner_preview.jsx [hunk 1]
```diff
@@ -78,8 +78,8 @@ export default class BannerPreview extends React.Component {
         currentDonation={0}
         balance={25.0}
         currentAmount={0}
-        bgImage={this.props.backgroundImage}
-        logo={this.props.logoImage}
+        bgImage={this.props.bannerImage.url}
```
Comment (yachtcaptain23) [inline app/javascript/packs/banner_preview.jsx]: What's the purpose of this change?

Comment (dlipeles) [reply app/javascript/packs/banner_preview.jsx]: In the editor I consolidated `logo` and `logoData` into one variable `logo`. On the preview side we get the url via `this.props.logo.url`. Same for bannerImage.

File: app/javascript/packs/banner_editor.jsx [hunk 2]
```diff
@@ -666,6 +856,43 @@ export default class BannerEditor extends React.Component {
...
       </div>
     );
+}
+
+  render() {
+
+    let component;
+    let that = this;
+
+    if(this.state.loading === true){
```
[the remark this answers is not in the document]
Comment (dlipeles) [reply app/javascript/packs/banner_editor.jsx]: Just created an item for refactoring at #1289 , included many of the points you added, thanks @petemill

File: app/javascript/packs/banner_editor.jsx [hunk 3]
```diff
@@ -97,6 +99,285 @@ export default class BannerEditor extends React.Component {
...
+        document.getElementsByClassName("brave-rewards-banner-container")[0].style.width = "600px"
+        document.getElementsByClassName("brave-rewards-banner-container")[0].style.height = "676px"
+        break;
+      case 'editor-logo-added':
+        this.setState({state: 'editor-logo-added'})
+        break;
+    }
+  }
+
+  renderControlBar(){
```
Comment (corymcdonald) [inline app/javascript/packs/banner_editor.jsx]: It makes more sense to me to extract these `renderX()` functions out into several components as these seem isolated. A large react file is hard for me to read 🙈

File: app/javascript/packs/banner_editor.jsx [hunk 4]
```diff
@@ -97,6 +99,285 @@ export default class BannerEditor extends React.Component {
...
+  }
+
+  handleDone(){
+
+    let that = this;
+    let logo = this.state.logo;
+    let bannerImage = this.state.bannerImage;
+    let img = this.editor.getImage();
+
+    switch(this.state.state){
```
Comment (corymcdonald) [inline app/javascript/packs/banner_editor.jsx]: can we rename the state variable `state` to be something different, i can see this getting confusing fast

File: app/javascript/packs/banner_editor.jsx (hunk 3 above)
Comment (dlipeles) [reply app/javascript/packs/banner_editor.jsx]: Agreed Cory, see the proposed refactor plan here #1289, it includes separate components 👍

File: app/javascript/packs/banner_editor.jsx (hunk 4 above)
Comment (dlipeles) [reply app/javascript/packs/banner_editor.jsx]: Any ideas? I chose state to represent the `state` of the banner since there will be several. `editing, logo-editing, banner-editing, preview, notification, etc.`

Comment (corymcdonald) [reply app/javascript/packs/banner_editor.jsx]: Hmmm, maybe something like `mode`?

Comment (dlipeles) [reply app/javascript/packs/banner_editor.jsx]: That's what I was thinking too, I really like state because of its [CS foundations](https://en.wikipedia.org/wiki/State_(computer_science))

But I'll decide on something in the refactor.

File: app/javascript/packs/banner_editor.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (petemill) [reply app/javascript/packs/banner_editor.jsx]: If you want to wait for multiple promises to resolve in parallel, you can use `Promise.all`, e.g.

```js
function fetchBanner() {
	return fetch(...).then(...)
}

function cropImage() {
	return new Promise(resolve => { img.onload = function() { ... resolve(image) ... }
}

Promise.all([
  fetchBanner(),
  cropImage()
])
.then([ fetchBannerResult, cropImageResult ], function () {
  this.setState(...)
})

```

[the remark this answers is not in the document]
Comment (dlipeles) [reply app/javascript/packs/banner_editor.jsx]: I think you'd be proud of this refactor Pete, the new fetchBanner()

```javascript
  async fetchBanner(){

    let that = this

    let url = '/publishers/' + document.getElementById("publisher_id").value + "/site_banners/fetch";
    let options = {
      method: 'GET',
      credentials: "same-origin",
      headers: {'Accept': 'text/html', 'X-Requested-With': 'XMLHttpRequest', 'X-CSRF-Token': document.head.querySelector("[name=csrf-token]").content}
    }

    let response = await fetch(url, options);
    let banner = await response.json();

    if(banner){
      // Scale images to fit banner editor
      let [logo, cover] = await Promise.all([that.scaleImage(banner.logoImage, 160, 160), that.scaleImage(banner.backgroundImage, 1200, 176)])

      that.setState({
          title: banner.title,
          description: banner.description,
          youtube: banner.socialLinks.youtube,
          twitter: banner.socialLinks.twitter,
          twitch: banner.socialLinks.twitch,
          donationAmounts: banner.donationAmounts,
          logo: {url: logo, data: null},
          cover: {url: cover, data: null},
          loading: false
        })}
    else{
      return;
    }

  }
```

File: app/javascript/packs/navbar.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    super(props);
+  }
+
+  render(){
+    return(
+      <Container>
+
+        <BrandBar mode={this.props.mode}>
+          <BrandImage src={DonationJar}/>
+          <BrandText>Tipping Banner</BrandText>
```
Comment (corymcdonald) [inline app/javascript/packs/navbar.jsx]: Let's ensure that we're internationalizing our javascript components

Comment (dlipeles) [reply app/javascript/packs/navbar.jsx]: Will make an item for i18n, it extends beyond this feature

**Comment by yachtcaptain23** [comment]
We also need to include either the npm lock file.

**Comment by dlipeles** [comment]
Merging to staging, @yachtcaptain23's feedback is implemented
